### ISSUE TITLE
Clipping gradient w.r.t. inputs at each time step for RNN/LSTM

### ISSUE THREAD

**Issue by asheshjain399** [tensorflow/tensorflow#2793]
New feature request:

While training LSTMs, is it often useful to clip the derivates w.r.t the inputs into the LSTM at each time step ([Alex Graves](http://arxiv.org/pdf/1308.0850v5.pdf), Sec 2.1). This is different from clipping the overall gradient. Theano supports this feature with `theano.gradient.grad_clip(tensor_var)`. Can we have a similar feature in tensorflow?

**Comment by ebrevdo** [comment]
It seems reasonable to add a clip_gradient op.  That said, you can always look through the output of tf.gradients(), find the gradient you want to clip, and pass it through a thresholding function before passing it to the optimizer.  That'll do what you want.  See ops/clip_ops.py to get a sense of how one may inspect the set of gradients and add modifications to some/all of them.

Marking as "contributions welcome" for the clip_gradient op.  Happy to review.

**Comment by Anmol6** [comment]
anyone working on this? If not, I'll be glad to work on it.

**Comment by DingKe** [comment]
any update?

**Comment by tfboyd** [comment]
No updates but we did just look at this issue during a big bug scrub.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by itsmeolivia** [comment]
@alextp do you know of a way to do this with weights?

**Comment by alextp** [comment]
It's easy to implement a clip_gradient "op" which acts like identity but clips the gradients using tfe.custom_gradient (which works normally in graph mode). Something like

```
@tfe.custom_gradient
def clip_gradient(x, clip):
  def grad(dresult):
    return [tf.clip_by_norm(dresult, clip)]
  return x, grad
```
